api/nakadi-event-bus-api.yaml

```diff
@@ -606,22 +488,394 @@ paths:
+          description: Client is not authenticated
+          schema:
+            $ref: '#/definitions/Problem'
+        '403':
+          description: Client is not authorized to perform this operation
+          schema:
+            $ref: '#/definitions/Problem'
+        '404':
+          description: Subscription not found
+          schema:
+            $ref: '#/definitions/Problem'
+        '500':
+          description: Server error
+          schema:
+            $ref: '#/definitions/Problem'
+        '503':
+          description: Service (temporarily) unavailable
+          schema:
+            $ref: '#/definitions/Problem'
+
+  /subscriptions/{subscription_id}/cursors:
+    get:
+      tags:
+        - managed-api
+      security:
+        - oauth2: ['nakadi.event_stream.read']
+      description: Exposes the currently committed offsets of a subscription.
+      parameters:
+        - $ref: '#/parameters/SubscriptionId'
+      responses:
+        '200':
+          description: Ok
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/Cursor'
+        '401':
+          description: Client is not authenticated
+          schema:
+            $ref: '#/definitions/Problem'
+        '403':
+          description: Client is not authorized to perform this operation
+          schema:
+            $ref: '#/definitions/Problem'
+        '404':
+          description: Subscription not found
+          schema:
+            $ref: '#/definitions/Problem'
+        '500':
+          description: Server error
+          schema:
+            $ref: '#/definitions/Problem'
+        '503':
+          description: Service (temporarily) unavailable
+          schema:
+            $ref: '#/definitions/Problem'
+    put:
+      tags:
+        - managed-api
+      security:
+        - oauth2: ['nakadi.event_stream.read']
+      description: |
+        Endpoint for committing offsets of the subscription. The behavior of commit is specific to `commit_mode`
+        parameter of the client stream (please see details in subscription streaming endpoint specification).
+      parameters:
+        - name: subscription_id
+          in: path
+          type: string
+          description: Id of subscription
+          required: true
+        - name: X-Nakadi-StreamId
+          in: header
+          type: string
+          description: |
+            Id of stream which client uses to read events. It is not possible to make a commit for a terminated or
+            none-existing stream. Also the client can't commit something which was not sent to his stream.
+          required: true
+        - name: cursors
+          in: body
+          description: Cursors with offsets to commit.
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/Cursor'
+          required: true
+      responses:
+        '200':
+          description: Offsets were committed
+        '207':
```

Comment:
@dehora thanks for explaining.

Being able to replay an event stream is an important point, and sending 409 in this case is surely the right solution.

Anyhow, I think that committing a running stream, handling a distributed rewind notification are two  different operations that should not been mixed up. Originally, I would have suggested to supported this by just opening up a new stream when replay is necessary, but with respect to resource usage and client orchestration I, can imagine to have a rewind operation on an existing managed stream too. In this case, I would suggest to extend the cursor with a `stream_version` that allows to distinguish whether a commit is requested for a stream after a rewind operation or might be a duplicate on the same stream version. This allows an outdated client with an invalid `stream_version` to be reset through a 409.

> So, I hope that helps. If the design is to perform a partial update, then I'd suggest POST plus 200 with a response body the client can examine. If there's a strong need to indicate at the response code level whether the update was complete or partial, then 204 for complete and 200 for partial with a response entity, ie plain HTTP, can support that without cherrypicking from and redefining WebDAV.

I absolutely agree with this suggestion, and in my opinion it matches the description below. This makes the standard use case very simple and straight forward to use for any client.